src/library/scala/collection/Iterator.scala

```diff
@@ -991,7 +991,8 @@ object Iterator extends IterableFactory[Iterator] {
     */
   def newBuilder[A]: Builder[A, Iterator[A]] =
     new ImmutableBuilder[A, Iterator[A]](empty[A]) {
-      override def addOne(elem: A): this.type = { elems = elems ++ single(elem); this }
+      override def addOne(elem: A): this.type             = { elems ++= single(elem); this }
+      override def addAll(xs: IterableOnce[A]): this.type = { elems ++= xs.iterator;  this }
```

Comment:
@dwijnand this implementation of `addAll` may have the side effect of changing laziness. I believe `++=` is lazy on Iterator, so now `addAll` will be lazy, but the inherited implementation is akin to `xs.foreach(addOne)` which will force. 

Maybe this whole Iterator builder business would be more efficient with 

```scala
def newBuilder[A]: Builder[A, Iterator[A]] = mutable.ArrayBuffer.newBuilder[A].mapResult(_.iterator)
```

Reply:
Or maybe the change to laziness is the intended point of this PR in that case then all is well :-)

Reply:
No, allocations was the point.  But I can't tell if the side change is bad or not...  Any opinions?

Reply:
I was going to say that it seems surprising that a builder would be lazy, because no other builders are, but of the lazy collections:
* LazyList.newBuilder is lazy
* View.newBuilder is strict 
* Stream (deprecated, RIP) is strict 

So there is mixed precedence, I think strict is still less surprising though.

Reply:
@scala/collections any opinions on whether we ought to give this some more thought...?

Reply:
I think lazyness is the right call here. It would feel pretty strange to me to create an `Iterator` and have it preemptively iterate through what was added. Or am I misunderstanding how `addAll` will be used?

Reply:
I agree that laziness is expected with iterators.

Reply:
I think that laziness is correct, even if it wasn't the intention of this PR. see also the discussion on #7573.

I didn't know that `View` had a `newBuilder` method; I think we should change it to be lazy (probably just `mapResult` on a `LazyList` builder). `Stream`'s builder should be lazy like `LazyList`'s, but it's largely unmaintained due to it being inferior to `LazyList` ;)

Reply:
Good link, thanks.

Reply:
Josh has now written on longer disquisition on this at https://github.com/scala/scala/pull/9483#issuecomment-776132800

Reply:
Since this is a "resolved" conversation and thus not very visible/discoverable, let's discuss further on #9483 rather than here.